File: ros2-basic/topic_simple/src/node_helloworld_py/node_helloworld_py/node_publisher_by_std_msgs.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class NodeHelloworld(Node):
    def __init__(self, name: str):
        super().__init__(name)
        self.count_ = 0
        # 创建 1 秒周期的定时器
        self.timer_ = self.create_timer(1.0, self.timer_callback)

        self.publisher_ = self.create_publisher(String, "helloworld_py_by_std_msgs", 10)

    def timer_callback(self):
        msg = String()
        msg.data = f"Hello World Python(std_msgs) Publisher:: {self.count_}"
        self.publisher_.publish(msg)
        self.get_logger().info(f"Hello World Python: {self.count_}")
        self.count_ += 1


def main(args=None):
    rclpy.init(args=args)
    node = NodeHelloworld("node_topic_publisher_helloworld_py_by_std_msgs")
    rclpy.spin(node)
    rclpy.shutdown()

File: ros2-basic/topic_simple/src/node_helloworld_cpp/src/node_publisher_by_std_msgs.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

class NodeHelloworld : public rclcpp::Node
{
public:
    NodeHelloworld(std::string name) : Node(name)
    {
        timer_ = this->create_wall_timer(
            std::chrono::seconds(1),
            std::bind(&NodeHelloworld::timer_callback, this));

        publisher_ = this->create_publisher<std_msgs::msg::String>("helloworld_cpp_by_std_msgs", 10);
    }

private:
    void timer_callback()
    {
        std_msgs::msg::String msg;
        msg.data = "Hello World C++(std_msgs) Publisher: " + std::to_string(count_);
        publisher_->publish(msg);
        RCLCPP_INFO(this->get_logger(), "Hello World Topic C++(std_msgs) Publisher: %d", ++count_);
    }

private:
    int count_;
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NodeHelloworld>("node_topic_publisher_helloworld_cpp_by_std_msgs");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}